### PR TITLE
adding AWSAuthenticator ported from internal 2.x branch

### PR DESCRIPTION
**Pull request by Nilithus**

### CODE REVIEW

**Comment by Nilithus** [comment]
@skluck

File: src/AWS/AWSAuthenticator.php [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+        return $this->aws->createS3($config);
+    }
+
+    /**
+     * @param string $region
+     * @param AWSStaticCredential|AWSRoleCredential|null $credential
+     *
+     * @return ServiceCatalogClient|null
+     */
+    public function getServiceCatalog($region, $credential)
```
Comment (skluck) [inline src/AWS/AWSAuthenticator.php]: Wont need this any time soon. can be removed.

File: src/AWS/AWSAuthenticator.php [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        return $this->aws->createCloudWatchLogs($config);
+    }
+
+    /**
+     * @param string $region
+     * @param AWSStaticCredential|AWSRoleCredential|null $credential
+     *
+     * @return array|null
+     */
+    public function getConfigWithCredentials($region, $credential)
```
Comment (skluck) [inline src/AWS/AWSAuthenticator.php]: I can't remember why this was public. Does it need to be?

File: src/AWS/CredentialProvider.php [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        $account = $credential->account();
+        $role = $credential->role();
+        $cacheKey = sprintf('%s_%s', $account, $role);
+
+        $assumeRoleProvider = AWSCredentialProvider::assumeRole([
+            'client' => $sts,
+            'assume_role_params' => [
+                'DurationSeconds' => $this->stsLifetime, # 900 - 3600
+                'RoleArn' => sprintf('arn:aws:iam::%s:%s', $account, $role),
+                'RoleSessionName' => 'hal-' . bin2hex(openssl_random_pseudo_bytes(4)), # unique identifier for auditing
```
Comment (skluck) [inline src/AWS/CredentialProvider.php]: We can use `\randombytes` now with php 7.

File: src/AWS/CredentialProvider.php [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        ];
+    }
+
+    /**
+     * @param AWSRoleCredential|null $credential
+     * @param string $region
+     *
+     * @return CredentialProvider|callable|null
+     */
+    public function getAssumeCredentials(AWSRoleCredential $credential, $region)
```
Comment (skluck) [inline src/AWS/CredentialProvider.php]: Should use `RoleCredentials` instead of `AssumeCredentials` now.

File: src/AWS/CredentialProvider.php [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    /**
+     * @param string $region
+     *
+     * @return StsClient|null
+     */
+    private function getStsWithInternalCredentials($region)
+    {
+        $credentials = $this->em
+            ->getRepository(Credential::class)
+            ->findOneBy(['isSpecial' => true, 'name' => $this->credentialName]);
```
Comment (skluck) [inline src/AWS/CredentialProvider.php]: `isSpecial` is `isInternal` now.

File: src/AWS/CredentialProvider.php [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+     * @return string
+     */
+    private function getSecret(AWSStaticCredential $credential)
+    {
+        if (!$secret = $credential->secret()) {
+            return '';
+        }
+
+        try {
+            $secret = $this->encryption->decrypt($secret);
```
Comment (skluck) [inline src/AWS/CredentialProvider.php]: We don't need to do try catch here, Crypto won't throw exceptions (or at least - we should guarantee it wont)

**Comment by skluck** [comment]
Ideally we'd encrypted sts credentials while cached, though we can let it slide for now. This will be refactored many times on the road to 3.0.

File: src/AWS/AWSAuthenticator.php (hunk 2 above)
Comment (Nilithus) [reply src/AWS/AWSAuthenticator.php]: Don't believe so -- will change to private.

**Comment by Nilithus** [comment]
@skluck changes made from code review

**Pull request merged**